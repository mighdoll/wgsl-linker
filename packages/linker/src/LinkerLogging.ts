import { ExtendedResult, ParserContext } from "../../mini-parse/src/Parser.js";
import { srcLog } from "../../mini-parse/src/ParserLogging.js";
import { logger } from "../../mini-parse/src/ParserTracing.js";
import { FoundRef } from "./TraverseRefs.js";

export function resultLog<T>(result: ExtendedResult<T>, ...msgs: any[]): void {
  srcLog(result.src, result.start, ...msgs);
}

export function refLog(ref: FoundRef, ...msgs: any[]): void {
  if (ref.kind !== "gen") {
    srcLog(ref.expMod.src, ref.elem.start, ...msgs);
  } else {
    logger(ref.name, ...msgs);
  }
}

export function ctxLog(ctx: ParserContext, ...msgs: any[]): void {
  srcLog(ctx.lexer.src, ctx.lexer.position(), ...msgs);
}