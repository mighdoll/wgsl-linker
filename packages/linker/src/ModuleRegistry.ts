import { linkWgslModule } from "./Linker.js";
import { parseModule, TextExport, TextModule } from "./ParseModule.js";
import { normalize, noSuffix, relativePath } from "./PathUtil.js";

/** A named function to transform code fragments (e.g. by inserting parameters) */
export interface Template {
  name: string;
  apply: ApplyTemplateFn;
}
export type CodeGenFn = (
  name: string,
  params: Record<string, string>
) => string;

export interface GeneratorExport {
  name: string;
  args: string[];
  generate: CodeGenFn;
}

/** a named code generation function */
export interface RegisterGenerator {
  /** export name for this generator */
  name: string;

  /** module namespace for this generator */
  moduleName: string;

  /** function to generate code at runtime */
  generate: CodeGenFn;

  /** arguments to pass when importing from this generator */
  args?: string[];
}

export type ApplyTemplateFn = (
  src: string,
  params: Record<string, any> // combination of external params and imp/exp params
) => string;

/** a single export from a module */
export type ModuleExport = TextModuleExport | GeneratorModuleExport;

export interface TextModuleExport {
  module: TextModule;
  export: TextExport;
  kind: "text";
}

export interface GeneratorModule {
  kind: "generator";
  name: string;
  exports: GeneratorExport[];
}

export interface GeneratorModuleExport {
  module: GeneratorModule;
  export: GeneratorExport;
  kind: "function";
}

/** unique index for naming otherwise unnamed generator modules and src files */
let unnamedCodeDex = 0;
let unnamedTextDex = 0;

export interface RegistryParams {
  /** record of file names an wgsl text for modules */
  wgsl?: Record<string, string>;

  /** alt interface to provide wgsl module texts w/o file names.
   * (usually best to provide filenames though, to allow relative imports) */
  rawWgsl?: string[];

  /** string template handlers for processing exported functions and structs */
  templates?: Template[];

  /** code generation functions */
  generators?: RegisterGenerator[];

  /** values for #if condition processing */
  conditions?: Record<string, any>;
}

/**
 * A ModuleRegistry collects exportable code fragments, code generator functions,
 * and template processors.
 *
 * The ModuleRegistry provides everything required for linkWgsl to process
 * #import statements and generate a complete wgsl shader.
 */
export class ModuleRegistry {
  // map from export names to a map of module names to exports
  private exports = new Map<string, ModuleExport[]>();
  private templates = new Map<string, ApplyTemplateFn>();
  private textModules: TextModule[] = [];
  private wgslSrc = new Map<string, string>();

  constructor(args?: RegistryParams) {
    if (!args) return;
    const { wgsl = {}, rawWgsl = [], templates = [], generators } = args;

    Object.entries(wgsl).forEach(([fileName, src]) =>
      this.wgslSrc.set(normalize(fileName), src)
    );
    rawWgsl.forEach((src) =>
      this.wgslSrc.set(`rawWgsl-${unnamedTextDex++}`, src)
    );
    templates && this.registerTemplate(...templates);
    generators?.map((g) => this.registerGenerator(g));
  }

  /**
   * Produce a linked wgsl string with all directives processed
   * (e.g. #import'd functions from other modules are inserted into the resulting string).
   * @param moduleName select the module to use as the root source
   * @param runtimeParams runtime parameters for #import/#export values,
   *  template values, and code generation parameters
   */
  link(moduleName: string, runtimeParams: Record<string, any> = {}): string {
    this._parseSrc(runtimeParams);
    const rootModule = this.findTextModule(moduleName);
    if (!rootModule) {
      console.error("no module found for ", moduleName);
      return "";
    }

    return linkWgslModule(rootModule, this, runtimeParams);
  }

  _parseSrc(runtimeParams: Record<string, any> = {}): void {
    this.textModules = [];
    this.wgslSrc.forEach((src, fileName) => {
      // dlog("parseSrc", { fileName, src });
      this.registerOneModule(src, runtimeParams, fileName);
    });
  }

  /** register one module's exports  */
  private registerOneModule(
    src: string,
    params: Record<string, any> = {},
    fileName: string,
    moduleName?: string
  ): void {
    const newFileName = fileName && normalize(fileName);
    const m = parseModule(src, newFileName, params, moduleName);
    this.addTextModule(m);
  }

  /** register a function that generates code on demand */
  registerGenerator(reg: RegisterGenerator): void {
    const exp: GeneratorExport = {
      name: reg.name,
      args: reg.args ?? [],
      generate: reg.generate
    };
    const module: GeneratorModule = {
      kind: "generator",
      name: reg.moduleName ?? `funModule${unnamedCodeDex++}`,
      exports: [exp]
    };
    const moduleExport: GeneratorModuleExport = {
      module,
      export: exp,
      kind: "function"
    };
    this.addModuleExport(moduleExport);
  }

  /** register a template processor  */
  registerTemplate(...templates: Template[]): void {
    templates.forEach((t) => this.templates.set(t.name, t.apply));
  }

  /** fetch a template processor */
  getTemplate(name: string): ApplyTemplateFn | undefined {
    return this.templates.get(name);
  }

  /** return a reference to an exported text fragment or code generator (i.e. in response to an #import request) */
  getModuleExport(
    requesting: TextModule,
    exportName: string,
    moduleName?: string
  ): ModuleExport | undefined {
    const exports = this.exports.get(exportName);
    if (!exports) {
      return undefined;
    } else if (moduleName?.startsWith(".")) {
      const searchName = relativePath(requesting.fileName, moduleName);
      const baseSearch = noSuffix(searchName);

      return exports.find((e) => {
        const fileName = (e.module as TextModule).fileName;
        if (!fileName) return false;
        if (fileName === searchName) return true;
        if (baseSearch === noSuffix(fileName)) return true;
      });
    } else if (moduleName) {
      return exports.find((e) => e.module.name === moduleName);
    } else if (exports.length === 1) {
      return exports[0];
    } else {
      const moduleNames = exports.map((e) => e.module.name).join(", ");
      console.warn(
        `Multiple modules export "${exportName}". (${moduleNames}) ` +
          `Use "#import ${exportName} from <moduleName>" to select which one import`
      );
    }
  }

  findTextModule(searchName: string): TextModule | undefined {
    const moduleNameMatch = this.textModules.find(
      (m) => m.name === searchName || m.fileName === searchName
    );
    // dlog({searchName, moduleNameMatch:moduleNameMatch?.name})
    // dlog({textModules:this.textModules.map(m => m.name)})
    if (moduleNameMatch) return moduleNameMatch;

    const baseSearch = normalize(searchName);
    const pathMatch = this.textModules.find(
      (m) => m.fileName === baseSearch || noSuffix(m.fileName) === baseSearch
    );
    if (pathMatch) return pathMatch;
  }

  private addTextModule(module: TextModule): void {
    this.textModules.push(module); // TODO dedupe?
    module.exports.forEach((e) => {
      const moduleExport: TextModuleExport = {
        module,
        export: e,
        kind: "text"
      };
      this.addModuleExport(moduleExport);
    });
  }

  private addModuleExport(moduleExport: ModuleExport): void {
    const expName = exportName(moduleExport);
    const existing = this.exports.get(expName);
    if (existing) {
      existing.push(moduleExport);
    } else {
      this.exports.set(expName, [moduleExport]);
    }
  }
}

function exportName(moduleExport: ModuleExport): string {
  if (moduleExport.kind === "text") {
    return moduleExport.export.ref.name;
  } else {
    return moduleExport.export.name;
  }
}
