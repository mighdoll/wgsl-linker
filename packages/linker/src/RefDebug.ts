import { dlog } from "berry-pretty";
import { FoundRef, TextRef } from "./TraverseRefs.js";
import { AbstractElem, CallElem } from "./AbstractElems.js";

export function printRef(r: FoundRef, msg = ""): void {
  const { kind, elem, rename } = r as TextRef;
  dlog(
    msg,
    {
      kind,
      rename,
    },
    elemToText("elem", elem)
  );
}

export function elemToText(msg: string, elem?: AbstractElem): string {
  if (!elem) return "";
  const { kind, ref: link, name = "" } = elem as CallElem;
  return `${msg}: {kind: ${kind}, name: ${name}, link: ${link !== undefined}}`;
}
