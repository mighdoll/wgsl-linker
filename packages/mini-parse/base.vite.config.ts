import { UserConfig } from "vite";
import tsconfigPaths from "vite-tsconfig-paths";
import dts from "vite-plugin-dts";
import { resolve } from "path";

export function baseViteConfig(): UserConfig {
  return {
    plugins: [
      tsconfigPaths(),
      dts(), // generate .d.ts files
    ],
    build: {
      lib: {
        name: "mini-parse",
        entry: [resolve(__dirname, "src/index.ts")],
      },
      minify: false,
      sourcemap: true,
    },
  };
}
