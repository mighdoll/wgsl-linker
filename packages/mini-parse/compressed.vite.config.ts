/// <reference types="vitest" />
import { LibraryOptions, defineConfig } from "vite";
// import { visualizer } from "rollup-plugin-visualizer";
import { baseViteConfig } from "./vite-base-config.js";

// Note that this will include the debug tracing code, so
// so the size estimate is an overestimate of production size

const config = baseViteConfig();
config.build!.emptyOutDir = false;
config.build!.minify = "terser";
(config.build!.lib as LibraryOptions).formats = ["es", "cjs"];
(config.build!.lib as LibraryOptions).fileName = "minified";

// config.plugins?.push(visualizer({ brotliSize: true, gzipSize: true })); // generate stats.html size report

export default defineConfig(config);
