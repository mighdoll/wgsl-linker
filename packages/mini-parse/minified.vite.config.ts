/// <reference types="vitest" />
import { LibraryOptions, defineConfig } from "vite";
import { baseViteConfig } from "./base.vite.config.js";
// import { visualizer } from "rollup-plugin-visualizer";

// Note that this will include the debug tracing code, so
// so the size estimate is an overestimate of production size

const config = baseViteConfig();
config.build!.emptyOutDir = false;
config.build!.minify = "terser";
(config.build!.lib as LibraryOptions).formats = ["es", "cjs"];
(config.build!.lib as LibraryOptions).fileName = "minified";

// config.plugins?.push(visualizer({ brotliSize: true, gzipSize: true })); // generate stats.html size report

export default defineConfig(config);
