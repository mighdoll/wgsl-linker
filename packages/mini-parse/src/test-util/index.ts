export * from "./LogCatcher.js";
export * from "./TestParse.js";
