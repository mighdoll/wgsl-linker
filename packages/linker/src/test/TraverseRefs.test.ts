import { _withBaseLogger } from "mini-parse";
import { logCatch } from "mini-parse/test-util";
import { expect, test } from "vitest";
import { ModuleRegistry } from "../ModuleRegistry.js";
import { FoundRef, TextRef, refName, traverseRefs } from "../TraverseRefs.js";
import { printRef } from "../RefDebug.js";
import { refLog } from "../LinkerLogging.js";
import { dlog } from "berry-pretty";

test("traverse a fn to struct ref", () => {
  const src = `
    #import AStruct 

    fn main() {
      let a:AStruct; 
    }
  `;
  const module1 = `
    #export
    struct AStruct {
      x: u32,
    }
  `;

  const refs = traverseTest(src, module1);
  const exp = refs[1] as TextRef;
  expect(exp.kind).eq("txt");
  expect(exp.elem.kind).eq("struct");
  expect(exp.elem.name).eq("AStruct");
});

test("traverse simple rust style import", () => {
  const main = `
    import bar::foo;
    module main
    fn main() { foo(); }
  `;
  const bar = `
    module bar
    export fn foo() { }
  `;
  const refs = traverseTest(main, bar);
  const exp = refs[1] as TextRef;
  expect(exp.kind).eq("txt");
  expect(exp.elem.kind).eq("fn");
  expect(exp.elem.name).eq("foo");
});

test("traverse nested import with params and support fn", () => {
  const src = `
    // #import foo(u32)
    fn bar() {
      foo(8u);
    }
  `;

  const module1 = `
    // #import zap
  
    // #export (A)
    fn foo(a: A) { 
      support(a);
      zap();
    }

    fn support() {}
  `;

  const module2 = `
    // #export 
    fn zap() {}
  `;

  const refs = traverseTest(src, module1, module2);
  const first = refs[1] as TextRef;
  const second = refs[2] as TextRef;
  expect(first.kind).toBe("txt");
  expect(first.expInfo?.expImpArgs).deep.eq([["A", "u32"]]);
  expect(second.kind).toBe("txt");
  expect(second.elem.name).eq("support");
});

test("traverse importing", () => {
  const src = `
    #import foo(A, B)
    fn main() {
      foo(k, l);
    } `;
  const module1 = `
    #export(C, D) importing bar(D)
    fn foo(c:C, d:D) { bar(d); } `;
  const module2 = `
    #export(X)
    fn bar(x:X) { } `;

  const refs = traverseTest(src, module1, module2);

  const importingRef = refs[2] as TextRef;
  expect(importingRef.expInfo?.expImpArgs).deep.eq([["X", "B"]]);
});

test("traverse double importing", () => {
  const src = `
    #import foo(A, B)
    fn main() {
      foo(k, l);
    } `;
  const module1 = `
    #export(C, D) importing bar(D)
    fn foo(c:C, d:D) { bar(d); } `;
  const module2 = `
    #export(X) importing zap(X)
    fn bar(x:X) { zap(x); } `;
  const module3 = `
    #export(Y) 
    fn zap(y:Y) { } `;

  const refs = traverseTest(src, module1, module2, module3);

  const expImpArgs = refs.flatMap((r) => {
    const er = r as TextRef;
    return er ? [er.expInfo?.expImpArgs] : [];
  });
  expect(expImpArgs[2]).deep.eq([["X", "B"]]);
  expect(expImpArgs[3]).deep.eq([["Y", "B"]]);
});

test("traverse importing from a support fn", () => {
  const src = `
    #import foo(A, B)
    fn main() {
      foo(k, l);
    } `;
  const module1 = `
    #export(C, D) importing support(D)
    fn foo(c:C, d:D) { support(d); } 
    
    #export(D) importing bar(D)
    fn support(d:D) { bar(d); }
    `;
  const module2 = `
    #export(X)
    fn bar(x:X) { } `;

  const refs = traverseTest(src, module1, module2);

  const expImpArgs = refs.flatMap((r) => {
    const er = r as TextRef;
    return er ? [{ name: er.elem.name, args: er.expInfo?.expImpArgs }] : [];
  });
  expect(expImpArgs).toMatchSnapshot();
});

test("traverse importing from a local call fails", () => {
  const src = `
    #import foo(A, B)
    fn main() {
      foo(k, l);
    } `;
  const module1 = `
    #export(C, D) importing bar(D)
    fn foo(c:C, d:D) { support(d); } 
    
    fn support(d:D) { bar(d); } //  need to mark this as an export with importing, so we can map params
    `;
  const module2 = `
    #export(X)
    fn bar(x:X) { } `;

  const { log } = traverseWithLog(src, module1, module2);
  expect(log.length).not.eq(0);
});

test("importing args don't match", () => {
  const src = `
    #import foo(A, B)
    fn main() {
      foo(k, l);
    } `;
  const module1 = `
    #export(C, D) importing bar(E)
    fn foo(c:C, d:D) { bar(d); } `;
  const module2 = `
    #export(X)
    fn bar(x:X) { } `;

  const { log } = traverseWithLog(src, module1, module2);

  expect(log).toMatchInlineSnapshot(`
    "importing arg doesn't match export  module: moduleFile0 moduleFile0
        #export(C, D) importing bar(E)   Ln 2
                                ^
    reference not found: X  module: moduleFile1 moduleFile1
        fn bar(x:X) { }    Ln 3
                 ^"
  `);
});

test("mismatched import export params", () => {
  const src = `
    #import foo(A, B)
    fn main() {
      foo(k, l);
    } `;
  const module1 = `
    #export(C) 
    fn foo(c:C) { } `;

  const { log } = traverseWithLog(src, module1);
  expect(log).toMatchInlineSnapshot(`
    "mismatched import and export params  module: main main
        #import foo(A, B)   Ln 2
        ^
     module: moduleFile0 moduleFile0
        #export(C)    Ln 2
        ^"
  `);
});

test("traverse var to rust style struct ref", () => {
  const main = `
     import foo::bar;
     module main
     var x: bar;
     fn main() { }
   `;
  const foo = `
      module foo
      export struct bar { f: f32 }
   `;

  const refs = traverseTest(main, foo);
  const structRef = refs.find(ref => ref.kind === "txt" && ref.elem.kind === "struct");
  expect(structRef).toBeDefined();
});

test("traverse a struct to struct ref", () => {
  const src = `
    #import AStruct 

    struct SrcStruct {
      a: AStruct,
    }
  `;
  const module1 = `
    #export
    struct AStruct {
      x: u32,
    }
  `;

  const refs = traverseTest(src, module1);
  expect(refs[1].kind).toBe("txt");
  expect(refName(refs[1])).toBe("AStruct");
});

test("traverse a global var to struct ref", () => {
  const src = `
    #import Uniforms

    @group(0) @binding(0) var<uniform> u: Uniforms;      
    `;
  const module1 = `
    #export
    struct Uniforms {
      model: mat4x4<f32>,
    }
  `;

  const refs = traverseTest(src, module1);
  const exp = refs[1] as TextRef;
  expect(exp.kind).eq("txt");
  expect(exp.elem.kind).eq("struct");
  expect(exp.elem.name).eq("Uniforms");
});

test("traverse transitive struct refs", () => {
  const src = `
    #import AStruct 

    struct SrcStruct {
      a: AStruct,
    }
  `;
  const module1 = `
    #import BStruct

    #export
    struct AStruct {
      s: BStruct,
    }
  `;
  const module2 = `
    #export
    struct BStruct {
      x: u32,
    }
  `;

  const refs = traverseTest(src, module1, module2);
  expect(refName(refs[1])).toBe("AStruct");
  expect(refName(refs[2])).toBe("BStruct");
});

test("traverse #export importing struct to struct", () => {
  const src = `
    #import AStruct(MyStruct)

    struct MyStuct {
      x: u32
    }

    struct HomeStruct {
      a:AStruct
    }
  `;
  const module1 = `
    #export(B) importing BStruct(B)
    struct AStruct {
      b: BStruct
    }
  `;

  const module2 = `
    #export(Y) 
    struct BStruct {
      Y: Y
    }
  `;
  const refs = traverseTest(src, module1, module2);
  expect(refName(refs[2])).toBe("AStruct");
  expect(refName(refs[3])).toBe("BStruct");
});

test("traverse ref from struct constructor", () => {
  const src = `
    #import AStruct

    fn main() {
      var x = AStruct(1u);
    }
  `;
  const module1 = `
    #export
    struct AStruct {
      b: u32
    }
  `;

  const refs = traverseTest(src, module1);
  expect(refName(refs[1])).toBe("AStruct");
});

test("traverse #extends", () => {
  const src = `
    #extends A
    struct B {
      x: u32
    }
  `;
  const module1 = `
    #export
    struct A {
      z: u32
    }
  `;
  const refs = traverseTest(src, module1);
  expect(refName(refs[1])).toBe("A");
});

test("traverse with local support struct", () => {
  const src = `
    #import A

    fn b() { var a: A; var b: B; }

    struct B { x: u32 }
  `;
  const module1 = `
    #export
    struct A { y: i32 }
  `;

  const refs = traverseTest(src, module1);
  const refNames = refs.map(refName);
  expect(refNames).deep.eq(["B", "b", "A", "B"]);
});

test("traverse from return type of function", () => {
  const src = `
    #import A

    fn b() -> A { }
  `;
  const module1 = `
    #export
    struct A { y: i32 }
  `;

  const refs = traverseTest(src, module1);
  const refNames = refs.map(refName);
  expect(refNames).deep.eq(["b", "A"]);
});

test("traverse skips built in fn and type", () => {
  const src = `
    fn foo() {
      bar();
      min(3,4);
      vec3(u);
    }
    fn bar() {}
  `;

  const { refs, log } = traverseWithLog(src);
  const refNames = refs.map(refName);
  // refs.map(r => refLog(r));
  expect(refNames).deep.eq(["foo", "bar", "bar"]); // TODO is this right?
  expect(log).eq("");
});

test("type inside fn with same name as fn", () => {
  // this will fail wgsl compilation, but as long as it doesn't hang the linker, we're ok
  const src = `
    fn foo() {
      let a:foo;
    }
    fn bar() {}
  `;
  const { refs, log } = traverseWithLog(src);
  expect(log).is.empty;
  expect(refs).length(3);
});

test("call inside fn with same name as fn", () => {
  const src = `
    fn foo() {
      foo();
    }
  `;
  const { refs, log } = traverseWithLog(src);
  expect(refs).length(1);
  expect(log).is.empty;
});

test("call cross reference", () => {
  const src = `
    fn foo() {
      bar();
    }

    fn bar() {
      foo();
    }
  `;
  const { refs, log } = traverseWithLog(src);
  const refNames = refs.map((r) => (r as TextRef).elem.name);
  expect(refNames).contains("foo");
  expect(refNames).contains("bar");
  expect(refNames).length(4); // TODO is this right?
  expect(log).is.empty;
});

test("struct self reference", () => {
  const src = `
    struct A {
      a: A,
      b: B,
    }
    struct B {
      f: f32,
    }
  `;
  const { log } = traverseWithLog(src);
  expect(log).is.empty;
});

test("struct cross reference", () => {
  const src = `
    struct A {
      b: B,
    }
    struct B {
      a: A,
    }
  `;
  const { refs, log } = traverseWithLog(src);
  expect(log).is.empty;
  const refNames = refs.map((r) => (r as any).elem.name);
  expect(refNames).includes("A");
  expect(refNames).includes("B");
  expect(refNames).length(4);
});

test("parse texture_storage_2d with texture format in type position", () => {
  const src = `var t: texture_storage_2d<rgba8unorm, write>;`;
  const { log } = traverseWithLog(src);
  expect(log).is.empty;
});

/** run traverseRefs with no filtering and return the refs and the error log output */
function traverseWithLog(
  src: string,
  ...modules: string[]
): { refs: FoundRef[]; log: string } {
  const { log, logged } = logCatch();
  const refs = _withBaseLogger(log, () => traverseTest(src, ...modules));

  return { refs, log: logged() };
}

/** run traverseRefs on the provided wgsl source strings
 * the first module is treated as the root
 */
function traverseTest(src: string, ...modules: string[]): FoundRef[] {
  const moduleFiles = Object.fromEntries(
    modules.map((m, i) => [`moduleFile${i}`, m])
  );
  const wgsl = { "./main": src, ...moduleFiles };
  const registry = new ModuleRegistry({ wgsl });
  const refs: FoundRef[] = [];
  const parsed = registry.parsed();
  const mainModule = parsed.findTextModule("./main")!;
  traverseRefs(mainModule, parsed, (ref) => {
    refs.push(ref);
    return true;
  });
  return refs;
}
