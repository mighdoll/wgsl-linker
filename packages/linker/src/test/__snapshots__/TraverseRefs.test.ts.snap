// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`traverse importing from a support fn 1`] = `
[
  {
    "args": undefined,
    "name": "main",
  },
  {
    "args": [
      [
        "C",
        "A",
      ],
      [
        "D",
        "B",
      ],
    ],
    "name": "foo",
  },
  {
    "args": [
      [
        "D",
        "B",
      ],
    ],
    "name": "support",
  },
  {
    "args": [
      [
        "X",
        "B",
      ],
    ],
    "name": "bar",
  },
]
`;
